### DISCUSSION TITLE
Allow `ICollection` to use list pattern matching, exclusively for `[]`

### DISCUSSION THREAD

**Discussion by RenderMichael** [dotnet/csharplang#7487] [Language Ideas]
To match a list with `[]`, all we really need is the length/count.

It would make things much more succinct to write `[]` instead of `{ Count: 0 }`.

edit: In fact, [the compiler](https://sharplab.io/#v2:CYLg1APgAgDABFAjAOgMIHsA2mCmBjAFwEt0A7AZwG4BYAKFgUQBYba6kA2OAI3SzgCyiABQBJADJFyBOJgCUcALwA+WXClwA2gF1WnHn0yCATGMnSAPOm4ArfAVXylj9eS266QA) doesn't use any indexing when matching on `[]`

**Comment by bernd5** [comment]
Good point.
There is a proposal about matching against [IEnumerable](https://github.com/dotnet/csharplang/blob/main/proposals/list-patterns-enumerables.md) but for "empty-collection-matching" countable should be sufficient.

**Reply by RenderMichael** [reply]
That's an interesting proposal. I see the LDM [commented on this](https://github.com/dotnet/csharplang/blob/13f8a815ea02c7f011d046b24256d9904fc19452/meetings/2021/LDM-2021-02-03.md#list-patterns-on-ienumerable) years ago with a not-immediately-dismissive tone.

However, that's a whole rabbit-hole requiring discussion, spec-ing, and more. I think the simple case I brought up here can be looked at without bringing in all the baggage from that proposal.

**Reply by 333fred** [reply]
We intended and still intend to support it, and @alrz has done a significant amount of work to bring it to life. I'm not especially enamored with the idea of shoehorning in support for `ICollection` here when it will come for free from something we're already actively investigating.

**Reply by CyrusNajmabadi** [reply]
that makes sense to me.

**Reply by alrz** [reply]
This could be an optimization in addition to TryGetNonEnumeratedCount. I think it's already mentioned in the proposal.

**Reply by RenderMichael** [reply]
I'm not entirely sure what the overlap is here. Presumably `enumerable is []` would look something like `enumerable.TryGetNonEnumeratedCount(out int count) ? count == 0 : enumerable.Any()`.

Whereas `collection is []` would do exactly what `list is []` does today, that is `(object)list != null && list.Count == 0`

**Comment by CyrusNajmabadi** [comment]
Possible candidate for an any-time change.  Thoughts @333fred ?

**Reply by 333fred** [reply]
I'm not convinced the special case is worth it, tbh. `IEnumerable` would cover it adequately.
